### PR TITLE
39 refactor to blockly generator

### LINKED ISSUE #39: Refactor generator into client Class and Service class
The Generator code should be refactored into BlocklyGenerator and BlocklyGeneratorService.

BlocklyGenerator is the client side class and should be responsible for connecting to the service and wrapping calls to the service (and contains the callback classes). This code is currently part of BlocklySectionsActivity but should be pulled out into its own class.

BlocklyGeneratorService is currently named CodeGeneratorService and shouldn't need many changes.

### PR DESCRIPTION
**Pull request by ChrisA89**
Fixes #39

---
This change is [Reviewable](https://reviewable.io/reviews/google/blockly-android/498)

### CODE REVIEW

File: blocklylib-core/src/main/java/com/google/blockly/android/WorkspaceFragment.java [hunk 1]
```diff
@@ -104,4 +104,8 @@ public Workspace getWorkspace() {
     public WorkspaceView getWorkspaceView() {
         return mWorkspaceView;
     }
+
+    public boolean hasBlocks() {
+        return getWorkspace().getRootBlocks().size() > 0;
```
Comment (ChrisA89) [inline blocklylib-core/src/main/java/com/google/blockly/android/WorkspaceFragment.java]: This is the logic that used to be called by onRunCode (above). I've extracted it here and it's now called in the BlocklyGenerator (below).

File: blocklylib-core/src/main/java/com/google/blockly/android/codegen/BlocklyGenerator.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     * Calls the Service to request code generation for the workspace passed in.
+     *
+     * @param workspaceFragment the workspace to request code generation for.
+     * @param codeGenerationCallback the callback to call with the generated code.
+     * @throws BlocklySerializerException if the workspace cannot be serialized to XML.
+     */
+    public void requestCodeGeneration(WorkspaceFragment workspaceFragment,
+                                      CodeGeneratorCallback codeGenerationCallback)
```
Comment (ChrisA89) [inline blocklylib-core/src/main/java/com/google/blockly/android/codegen/BlocklyGenerator.java]: The callback could be passed in the constructor, but I think it makes the outcome of the method clearer if passed in here.

File: blocklylib-core/src/main/java/com/google/blockly/android/codegen/BlocklyGenerator.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+                mGeneratorService.requestCodeGeneration(
+                    new CodeGenerationRequest(
+                        serialized.toString(),
+                        codeGenerationCallback,
+                        mBlockDefinitionsJsonPaths,
+                        mGeneratorsJsPaths)
+                );
+            } else {
+                Log.i(TAG, "Generator not bound to service. Skipping run request.");
```
Comment (ChrisA89) [inline blocklylib-core/src/main/java/com/google/blockly/android/codegen/BlocklyGenerator.java]: This didn't used to be logged, but I think it's useful.

**Comment by AnmAtAnm** [comment]
For the most part, this looks great. A few easy issues noted below.

I think we should keep the term CodeGeneratorService (or at least BlocklyCodeGeneratorService if you really think it needs "Blockly" in the name, a fair point for something listed in the manifest). [Code generator](https://developers.google.com/blockly/guides/configure/web/code-generators) is the term we use in the docs to describe the JavaScript functions this runs.

Relatedly, in my mind, "Blockly generator" invokes the idea of those JavaScript functions. I think the BlocklyGenerator you have below is really a helper or manager for the code generation service. Hence, CodeGenerationManager or ...Helper is probably a better name.

Thanks a bunch!!

---

Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 308 at r1](https://reviewable.io:443/reviews/google/blockly-android/498#-KacEyXWLfyFsw9wRGAn:-KacEyXWLfyFsw9wRGAo:bnt2om5) ([raw file](https://github.com/google/blockly-android/blob/c089a61e0ef839d110be60ff1587db99d9c244db/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L308)):*
> ```Java
>         mClipDataHelper = onCreateClipDataHelper();
>         mBlocklyGenerator = new BlocklyGenerator(
>             getBlockDefinitionsJsonPaths(), getGeneratorsJsPaths());
> ```

`getBlockDefinitionsJsonPaths()` and `getGeneratorsJsPaths()` should be called and passed in during `onRunCode()`, so an app can change the available block set or output language based on other parts of the app (e.g., a output language selector).

You probably want to pass in the `this` Context for the below onResume()/onPause() suggestions.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 429 at r1](https://reviewable.io:443/reviews/google/blockly-android/498#-KacFeFPePwurH8uUSbd:-Kac4YVmZx9H-tNNSgv3:b54up6q) ([raw file](https://github.com/google/blockly-android/blob/c089a61e0ef839d110be60ff1587db99d9c244db/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L429)):*
> ```Java
>         super.onResume();
>         Intent intent = new Intent(this, BlocklyGeneratorService.class);
>         bindService(intent, mBlocklyGenerator.getCodeGenerationService(), Context.BIND_AUTO_CREATE);
> ```

I suggest wrapping these two lines in CodeGenerationManager.onResume().  This, and the matching onPause() below, should complete the transition to a higher level API, with no direct references to the Service.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 456 at r1](https://reviewable.io:443/reviews/google/blockly-android/498#-KacG4y_N3azs8GdPg4k:-KacG4y_N3azs8GdPg4l:bj6npou) ([raw file](https://github.com/google/blockly-android/blob/c089a61e0ef839d110be60ff1587db99d9c244db/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L456)):*
> ```Java
>     protected void onPause() {
>         super.onPause();
>         unbindService(mBlocklyGenerator.getCodeGenerationService());
> ```

And this line could be moved to CodeGenerationManager.onPause().

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 765 at r1](https://reviewable.io:443/reviews/google/blockly-android/498#-KacIYKtGAammaE1IsJo:-KacIYKtGAammaE1IsJp:b-ug9o0r) ([raw file](https://github.com/google/blockly-android/blob/c089a61e0ef839d110be60ff1587db99d9c244db/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L765)):*
> ```Java
> 
>     /**
>      * Runs the code generator. Called when user selects "Run" action.
> ```

Please add:
`Gets the latest block definitions and generator code by calling {@link #getBlockDefinitionsJsonPaths()} and {@link #getGeneratorsJsPaths()} just before invoking generation.`

---

*[blocklylib-core/src/main/java/com/google/blockly/android/WorkspaceFragment.java, line 109 at r1](https://reviewable.io:443/reviews/google/blockly-android/498#-KaS_0kdIMWztxkKJS9d-r1-109:-Kac5E92JhCVv9V_xIi9:b-9qj12b) ([raw file](https://github.com/google/blockly-android/blob/c089a61e0ef839d110be60ff1587db99d9c244db/blocklylib-core/src/main/java/com/google/blockly/android/WorkspaceFragment.java#L109)):*
<details><summary><i>Previously, ChrisA89 wrote…</i></summary><blockquote>

This is the logic that used to be called by onRunCode (above). I've extracted it here and it's now called in the BlocklyGenerator (below).
</blockquote></details>

Move this into Workspace, instead of WorkspaceFragement. Add a JavaDoc comment when you do.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/codegen/BlocklyGenerator.java, line 18 at r1](https://reviewable.io:443/reviews/google/blockly-android/498#-KacEUFT1_0d6ZxiB9Ku:-KacEUFT1_0d6ZxiB9Kv:bgyv8d6) ([raw file](https://github.com/google/blockly-android/blob/c089a61e0ef839d110be60ff1587db99d9c244db/blocklylib-core/src/main/java/com/google/blockly/android/codegen/BlocklyGenerator.java#L18)):*
> ```Java
>  * Client Side class responsible for connecting to the {@link BlocklyGeneratorService}.
>  */
> public class BlocklyGenerator {
> ```

CodeGenerationHelper or Manager. Reasoning detailed at the top of this review.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/codegen/BlocklyGenerator.java, line 46 at r1](https://reviewable.io:443/reviews/google/blockly-android/498#-Kac6jbmw7D_JGtn6GCS:-Kac6jbnevCLW_9CVX2e:bcgbf1n) ([raw file](https://github.com/google/blockly-android/blob/c089a61e0ef839d110be60ff1587db99d9c244db/blocklylib-core/src/main/java/com/google/blockly/android/codegen/BlocklyGenerator.java#L46)):*
> ```Java
>     }
> 
>     public ServiceConnection getCodeGenerationService() {
> ```

Probably don't need this if we implement the onResume()/onPause() methods suggested above.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/codegen/BlocklyGenerator.java, line 58 at r1](https://reviewable.io:443/reviews/google/blockly-android/498#-KaS_0kdIMWztxkKJS9e-r1-58:-Kac5Iud1nYIDJRXyQs7:b-a1x3tm) ([raw file](https://github.com/google/blockly-android/blob/c089a61e0ef839d110be60ff1587db99d9c244db/blocklylib-core/src/main/java/com/google/blockly/android/codegen/BlocklyGenerator.java#L58)):*
<details><summary><i>Previously, ChrisA89 wrote…</i></summary><blockquote>

The callback could be passed in the constructor, but I think it makes the outcome of the method clearer if passed in here.
</blockquote></details>

The first argument should be a Workspace, since there is nothing particular about the UI/Fragment that is needed here.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/codegen/BlocklyGenerator.java, line 74 at r1](https://reviewable.io:443/reviews/google/blockly-android/498#-KaS_0kdIMWztxkKJS9e-r1-74:-Kac5ZajeNeCOVEcgEbt:bn1wz0t) ([raw file](https://github.com/google/blockly-android/blob/c089a61e0ef839d110be60ff1587db99d9c244db/blocklylib-core/src/main/java/com/google/blockly/android/codegen/BlocklyGenerator.java#L74)):*
<details><summary><i>Previously, ChrisA89 wrote…</i></summary><blockquote>

This didn't used to be logged, but I think it's useful.
</blockquote></details>

:thumbsup:

---

*[blocklylib-core/src/main/java/com/google/blockly/android/codegen/BlocklyGeneratorService.java, line 50 at r1](https://reviewable.io:443/reviews/google/blockly-android/498#-KacEjW_JR1fcLvYzGIZ:-KacEjW_JR1fcLvYzGI_:b-leyd5z) ([raw file](https://github.com/google/blockly-android/blob/c089a61e0ef839d110be60ff1587db99d9c244db/blocklylib-core/src/main/java/com/google/blockly/android/codegen/BlocklyGeneratorService.java#L50)):*
> ```Java
>  * to generate code.
>  */
> public class BlocklyGeneratorService extends Service {
> ```

Revert rename.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/498#-:-Kac7sNHCkJV3uETHj0v:b-y28n9e)*

**Comment by ChrisA89** [comment]
I'll sort this out now. The rename was mainly because issue #39 asked for it.

**Comment by AnmAtAnm** [comment]
Oh, looky there. Huh... Sorry for the confusion. Should be easy to fix. Thanks!

**Comment by AnmAtAnm** [comment]
:LGMT:

**Pull request merged**